Problem Statement

Your client is a large Multinational Corporation, and they have nine broad verticals
across the organization. One of the problems your client faces is identifying the right
people for promotion (only for the manager position and below) and preparing them in
time.

Currently the process, they are following is:

● They first identify a set of employees based on recommendations/ past
performance.

● Selected employees go through the separate training and evaluation program for each vertical.

● These programs are based on the required skill of each vertical. At the end of the program, based on various factors such as training performance, KPI completion (only employees with KPIs completed greater than 60% are considered) etc., the employee gets a promotion.

For the process mentioned above, the final promotions are only announced after the evaluation, and this leads to a delay in transition to their new roles. Hence, the company needs your help in identifying the eligible candidates at a particular checkpoint so that they can expedite the entire promotion cycle.

They have provided multiple attributes around employees’ past and current performance along with demographics. Now, The task is to predict whether a potential promotee at a checkpoint will be promoted or not after the evaluation process.

Dataset

● Dataset URL: https://bit.ly/2ODZvLCHRDataset

● Glossary URL: https://bit.ly/2Wz3sWcGlossary

Project Source: https://bit.ly/2CFzoRX

In [11]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

data = pd.read_csv('https://bit.ly/2ODZvLCHRDataset')

data.head(5)

employee_id         department     region         education gender  \
0        65438  Sales & Marketing   region_7  Master's & above      f   
1        65141         Operations  region_22        Bachelor's      m   
2         7513  Sales & Marketing  region_19        Bachelor's      m   
3         2542  Sales & Marketing  region_23        Bachelor's      m   
4        48945         Technology  region_26        Bachelor's      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   35                   5.0   
1               other                1   30                   5.0   
2            sourcing                1   34                   3.0   
3               other                2   39                   1.0   
4               other                1   45                   3.0   

   length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                  8              1            0                  49   
1                  4              0            0                  60   
2                  7              0            0                  50   
3                 10              0            0                  50   
4                  2              0            0                  73   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0

In [10]:
data.shape

(54808, 14)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54808 entries, 0 to 54807
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   employee_id           54808 non-null  int64  
 1   department            54808 non-null  object 
 2   region                54808 non-null  object 
 3   education             52399 non-null  object 
 4   gender                54808 non-null  object 
 5   recruitment_channel   54808 non-null  object 
 6   no_of_trainings       54808 non-null  int64  
 7   age                   54808 non-null  int64  
 8   previous_year_rating  50684 non-null  float64
 9   length_of_service     54808 non-null  int64  
 10  KPIs_met >80%         54808 non-null  int64  
 11  awards_won?           54808 non-null  int64  
 12  avg_training_score    54808 non-null  int64  
 13  is_promoted           54808 non-null  int64  
dtypes: float64(1), int64(8), object(5)
memory usage: 5.9+ MB


In [12]:
# Get total missing values in 'previous_year_rating' column
data.previous_year_rating.isnull().sum()

4124

In [17]:
def fill_missing(df, column, value):
    df_copy = df.copy()
    df_copy[column].fillna(value=value,inplace=True)
    return df_copy
    
data = fill_missing(data, "previous_year_rating", 0.0)

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54808 entries, 0 to 54807
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   employee_id           54808 non-null  int64  
 1   department            54808 non-null  object 
 2   region                54808 non-null  object 
 3   education             52399 non-null  object 
 4   gender                54808 non-null  object 
 5   recruitment_channel   54808 non-null  object 
 6   no_of_trainings       54808 non-null  int64  
 7   age                   54808 non-null  int64  
 8   previous_year_rating  54808 non-null  float64
 9   length_of_service     54808 non-null  int64  
 10  KPIs_met >80%         54808 non-null  int64  
 11  awards_won?           54808 non-null  int64  
 12  avg_training_score    54808 non-null  int64  
 13  is_promoted           54808 non-null  int64  
dtypes: float64(1), int64(8), object(5)
memory usage: 5.9+ MB


In [18]:
data['promoted'] = np.where(data['is_promoted'] == 1, 'Yes', 'No')
data

employee_id         department     region         education gender  \
0            65438  Sales & Marketing   region_7  Master's & above      f   
1            65141         Operations  region_22        Bachelor's      m   
2             7513  Sales & Marketing  region_19        Bachelor's      m   
3             2542  Sales & Marketing  region_23        Bachelor's      m   
4            48945         Technology  region_26        Bachelor's      m   
...            ...                ...        ...               ...    ...   
54803         3030         Technology  region_14        Bachelor's      m   
54804        74592         Operations  region_27  Master's & above      f   
54805        13918          Analytics   region_1        Bachelor's      m   
54806        13614  Sales & Marketing   region_9               NaN      m   
54807        51526                 HR  region_22        Bachelor's      m   

      recruitment_channel  no_of_trainings  age  previous_year_rating  \
0                sourcing                1   35                   5.0   
1                   other                1   30                   5.0   
2                sourcing                1   34                   3.0   
3                   other                2   39                   1.0   
4                   other                1   45                   3.0   
...                   ...              ...  ...                   ...   
54803            sourcing                1   48                   3.0   
54804               other                1   37                   2.0   
54805               other                1   27                   5.0   
54806            sourcing                1   29                   1.0   
54807               other                1   27                   1.0   

       length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                      8              1            0                  49   
1                      4              0            0                  60   
2                      7              0            0                  50   
3                     10              0            0                  50   
4                      2              0            0                  73   
...                  ...            ...          ...                 ...   
54803                 17              0            0                  78   
54804                  6              0            0                  56   
54805                  3              1            0                  79   
54806                  2              0            0                  45   
54807                  5              0            0                  49   

       is_promoted promoted  
0                0       No  
1                0       No  
2                0       No  
3                0       No  
4                0       No  
...            ...      ...  
54803            0       No  
54804            0       No  
54805            0       No  
54806            0       No  
54807            0       No  

[54808 rows x 15 columns]

In [29]:
 #assign the categorical data into cats object
cats = ['department','region','education','gender','recruitment_channel']

# assign the numerical data into nums object
nums = ['employee_id','no_of_trainings','age','previous_year_rating','length_of_service','KPIs_met >80%','awards_won?',
        'avg_training_score','is_promoted']

In [21]:
#Statistical Categorical Data

# assign the categorical data into cats object
cats = ['department','region','education','gender','recruitment_channel']

# assign the numerical data into nums object
nums = ['employee_id','no_of_trainings','age','previous_year_rating','length_of_service','KPIs_met >80%','awards_won?',
        'avg_training_score','is_promoted']

data[cats].describe()


department    region   education gender recruitment_channel
count               54808     54808       52399  54808               54808
unique                  9        34           3      2                   3
top     Sales & Marketing  region_2  Bachelor's      m               other
freq                16840     12343       36669  38496               30446

Conclusion from Statistical Categorical Data

Data majority is m (Male) with frequency 38496
Mostly work in Sales & Marketing departments
Mostly education degree is Bachelor's

In [31]:
# check the cardinality or unique value from categorical data 
# and assign thgem into data_cat_unique object
display(data.select_dtypes(include=['object']).columns)
print(data.select_dtypes(include=object).shape)
data_cat = data.select_dtypes(include=['object'])
data_cat.head(3)

Index(['department', 'region', 'education', 'gender', 'recruitment_channel',
       'promoted'],
      dtype='object')

(54808, 6)


department     region         education gender recruitment_channel  \
0  Sales & Marketing   region_7  Master's & above      f            sourcing   
1         Operations  region_22        Bachelor's      m               other   
2  Sales & Marketing  region_19        Bachelor's      m            sourcing   

  promoted  
0       No  
1       No  
2       No

In [32]:
data_cat_unique = data_cat.nunique().reset_index()
data_cat_unique.columns = ['feature', 'unique value']
data_cat_unique = data_cat_unique.sort_values('unique value', ascending=False).reset_index(drop=True)
data_cat_unique

feature  unique value
0               region            34
1           department             9
2            education             3
3  recruitment_channel             3
4               gender             2
5             promoted             2

There are 5 features with categorical data
- region has 34 unique value
- department has 9 unique value
- education has 3 unique value
- recruitment_channel has 3 unique value
- gender has 2 unique value

In [33]:
#Probability to Get Promotion based on KPIs_met >80%

kpi_probs= data.groupby(['KPIs_met >80%', 'is_promoted']).agg({
    'department': ['count']
}).reset_index()

kpi_probs.columns = ['KPIs_met >80%', 'is_promoted', 'employees']

kpi_probs = pd.pivot_table(kpi_probs, 
               index = 'KPIs_met >80%',
               columns = 'is_promoted',
               values = 'employees'
              ).reset_index()
kpi_probs.columns = ['KPIs_met >80%','not_promoted','promoted']

kpi_probs['total_employees'] = kpi_probs['not_promoted']+kpi_probs['promoted']

kpi_probs['promotion_probs'] = (kpi_probs['promoted']/(kpi_probs['not_promoted']+kpi_probs['promoted']))*100

kpi_probs = kpi_probs.sort_values('promotion_probs', ascending=False).reset_index(drop=True)

kpi_probs['KPIs_met >80%'] = np.where(kpi_probs['KPIs_met >80%'] == 1, 'YES', "NO")

kpi_probs

KPIs_met >80%  not_promoted  promoted  total_employees  promotion_probs
0           YES         16029      3262            19291        16.909440
1            NO         34111      1406            35517         3.958668

employees with KPI> 80% have higher probabilities than others:
- employees met KPI > 80% have 16.91% probabilities (3262 employees) to get promotion
- who didn't met KPI > 80% only have 3.96% probablities (1406 employees) to get promotio

In [34]:
#Probability to Get Promotion based on previous_year_rating

rating_probs = data.groupby(['previous_year_rating','is_promoted']).agg({
    'department': ['count']
}).reset_index()

rating_probs.columns = ['previous_year_rating', 'is_promoted', "employees"]

rating_probs = pd.pivot_table(rating_probs,
                              index = 'previous_year_rating',
                              columns = 'is_promoted',
                              values = "employees"
                             ).reset_index()

rating_probs.columns = ['previous_year_rating', 'not_promoted', 'promoted']
rating_probs['total_employees'] = rating_probs['not_promoted'] + rating_probs['promoted']
rating_probs['promotion_probs'] = (rating_probs['promoted']/(rating_probs['not_promoted']+rating_probs['promoted']))*100
rating_probs = rating_probs.sort_values('promotion_probs', ascending=False).reset_index(drop=True)
rating_probs

previous_year_rating  not_promoted  promoted  total_employees  \
0                   5.0          9820      1921            11741   
1                   0.0          3785       339             4124   
2                   4.0          9093       784             9877   
3                   3.0         17263      1355            18618   
4                   2.0          4044       181             4225   
5                   1.0          6135        88             6223   

   promotion_probs  
0        16.361468  
1         8.220175  
2         7.937633  
3         7.277903  
4         4.284024  
5         1.414109

Employee with rating 5 has higher probabilities
to get promotion, about 16.36% (1921 employees) 
from total 11741 employees with rating 5

In [35]:
# Probability to Get Promotion based on awards_won?
awards_probs = data.groupby(['awards_won?', "is_promoted"]).agg({
    'department': ['count']
}).reset_index()

awards_probs.columns = ["awards_won?", "is_promoted", "employees"]

awards_probs = pd.pivot_table(awards_probs,
                              index = "awards_won?",
                              columns = "is_promoted",
                              values = "employees"
                             ).reset_index()

awards_probs.columns = ["awards_won?", "not_promoted", "promoted"]
awards_probs['total_employees'] = awards_probs['not_promoted'] + awards_probs['promoted']
awards_probs["promotion_probs"] = (awards_probs['promoted']/(awards_probs["not_promoted"]+awards_probs["promoted"]))*100
awards_probs = awards_probs.sort_values("promotion_probs", ascending=False).reset_index(drop=True)
awards_probs['awards_won?'] = np.where(awards_probs['awards_won?'] == 1,
                                      'YES', 'NO')
awards_probs

awards_won?  not_promoted  promoted  total_employees  promotion_probs
0         YES           711       559             1270        44.015748
1          NO         49429      4109            53538         7.674922

Employees who won any awards in the previous year
have higher probabilities to get promotion, 
it's aobut 44.02% from total 1270 who won awards

In [38]:
# Probability to Get Promotion based on no_of_trainings

nof_probs = data.groupby(['no_of_trainings', 'is_promoted']).agg({
    'department': ['count']
}).reset_index()

nof_probs.columns = ['no_of_trainings', 'is_promoted','employees']

nof_probs = pd.pivot_table(nof_probs,
                             index = 'no_of_trainings',
                             columns = 'is_promoted',
                             values = 'employees').reset_index()

nof_probs.columns = ['no_of_trainings','not_promoted','promoted']
nof_probs['total_employees'] = nof_probs['not_promoted'] + nof_probs['promoted']
nof_probs['promotion_probs'] = (nof_probs['promoted']/(nof_probs['not_promoted']+nof_probs['promoted']))*100
nof_probs = nof_probs.sort_values('promotion_probs', ascending=False).reset_index(drop=True)
nof_probs

no_of_trainings  not_promoted  promoted  total_employees  promotion_probs
0                1       40468.0    3910.0          44378.0         8.810672
1                2        7382.0     605.0           7987.0         7.574809
2                3        1654.0     122.0           1776.0         6.869369
3                4         442.0      26.0            468.0         5.555556
4                6          42.0       2.0             44.0         4.545455
5                5         125.0       3.0            128.0         2.343750
6                7          12.0       NaN              NaN              NaN
7                8           5.0       NaN              NaN              NaN
8                9           5.0       NaN              NaN              NaN
9               10           5.0       NaN              NaN              NaN

In [39]:
# Drop no value columns
nof_probs = nof_probs.drop(nof_probs[nof_probs['no_of_trainings']>= 7].index)
nof_probs

no_of_trainings  not_promoted  promoted  total_employees  promotion_probs
0                1       40468.0    3910.0          44378.0         8.810672
1                2        7382.0     605.0           7987.0         7.574809
2                3        1654.0     122.0           1776.0         6.869369
3                4         442.0      26.0            468.0         5.555556
4                6          42.0       2.0             44.0         4.545455
5                5         125.0       3.0            128.0         2.343750

Most of employees only take 1 training.Training affects the probabilities for promoted employees.
In other words, the probabilities is higher for employees who had 1 training